### ISSUE TITLE
[Barreleye latest] Failed building obmc-phosphor-image due to phosphor-ipmi-fru compiling fails

### ISSUE THREAD

**Issue by yadron** [openbmc/openbmc#1441]
**"tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/build/fru-gen.cpp" broken for Barreleye currently, so please fix generation of fru-gen.cpp scripts or/and used configuration of Barreleye.**

First it looked for me like gcc updated from 5.3.0 to 6.2.0 and as result failed after that (or maybe the gcc itself Ok, but new version of scripts for fru-gen failed):

```
ERROR: phosphor-ipmi-fru-1.0-r1 do_compile: oe_runmake failed
ERROR: phosphor-ipmi-fru-1.0-r1 do_compile: Function failed: do_compile (log file is located at /home/build/openbmc3/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/temp/log.do_compile.24703)
ERROR: Logfile of failure stored in: /home/build/openbmc3/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/temp/log.do_compile.24703
```

The actual error message is (checked twice):

cat /home/build/openbmc3/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/temp/log.do_compile.14728
```
arm-openbmc-linux-gnueabi-libtool: compile:  arm-openbmc-linux-gnueabi-g++ -march=armv5e -marm --sysroot=/home/build/openbmc3/openbmc/build/tmp/sysroots/barreleye -DHAVE_CONFIG_H -I. -I../git -O2 -pipe -g -feliminate-unused-debug-types -fdebug-prefix-map=/home/build/openbmc3/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1=/usr/src/debug/phosphor-ipmi-fru/1.0-r1 -fdebug-prefix-map=/home/build/openbmc3/openbmc/build/tmp/sysroots/x86_64-linux= -fdebug-prefix-map=/home/build/openbmc3/openbmc/build/tmp/sysroots/barreleye= -fvisibility-inlines-hidden -flto -fPIC -Wall -Werror -c ../git/frup.cpp  -fPIC -DPIC -o .libs/libwritefrudata_la-frup.o
arm-openbmc-linux-gnueabi-g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
Makefile:648: recipe for target 'libwritefrudata_la-fru-gen.lo' failed
make[1]: *** [libwritefrudata_la-fru-gen.lo] Error 1
make[1]: Leaving directory '/home/build/openbmc3/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/build'
Makefile:416: recipe for target 'all' failed
make: *** [all] Error 2
ERROR: Function failed: do_compile (log file is located at /home/build/openbmc3/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/temp/log.do_compile.14728)
```

(see openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/build during building)

Looks like this line of Makefile failed:

```
648         $(AM_V_CXX)$(LIBTOOL) $(AM_V_lt) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=compile $(CXX) $(DEFS) $(DEFAULT_INCLUDES)      $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libwritefrudata_la_CXXFLAGS) $(CXXFLAGS) -c -o libwritefrudata_la-fru-gen.lo `test -f 'fru-ge     n.cpp' || echo '$(srcdir)/'`fru-gen.cpp
```

Probably the problem with auto-generated file
openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/build/fru-gen.cpp

It was empty in previous versions of OpenBmc (like

```
cat openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/build/fru-gen.cpp
// !!! WARNING: This is a GENERATED Code..Please do NOT Edit !!!
#include <iostream>
#include "frup.hpp"

extern const FruMap frus = {
};
```
), and now filled with content:

cat build/fru-gen.cpp |more
```
// !!! WARNING: This is a GENERATED Code..Please do NOT Edit !!!
#include <iostream>
#include "frup.hpp"

extern const FruMap frus = {
   {1,{

         {"/system/chassis/motherboard/cpu0",{
             {"xyz.openbmc_project.Inventory.Decorator.Asset",{
                 {"PartNumber",{
                     {"IPMIFruSection","Board"},
                     {"IPMIFruProperty","Part Number"},
                 }},
...
```

If I overwrite filled file with empty one (like copying prev version above current
`cp /home/build/openbmc2/openbmc/build/tmp/work/armv5e-openbmc-linux-gnueabi/phosphor-ipmi-fru/1.0-r1/build/fru-gen.cpp ../build/fru-gen.cpp
`) and build again (bitbake obmc-phosphor-image), then everything compiled well (e.g. "Attempted 3298 tasks of which 3207 didn't need to be rerun and all succeeded." and
```
build@obmc build]$ ls -l tmp/deploy/images/barreleye/flash-barreleye-20170406191329 
-rw-rw-r--. 1 build build 33554432 Apr  6 22:15 tmp/deploy/images/barreleye/flash-barreleye-20170406191329
```
).

**As result: "build/fru-gen.cpp" broken for Barreleye currently, so please fix generation of fru-gen.cpp scripts or/and used configuration of Barreleye.**

**Comment by mine260309** [comment]
MSI building P8 machine hit the same issue, and I confirm it can be workaround by increasing the RAM, either physical RAM or swap space would help.
On MSI's case, it is consuming more than 20GiB RAM building it.

So this **is** the same issue caused by GCC6: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80290

**Comment by rfrandse** [comment]
<https://gerrit.openbmc-project.xyz/4544> fru_gen: reduce map in generated data
Resolves  #1441 [Barreleye latest] Failed building obmc-phosphor-image due to phosphor-ipmi-fru compiling fails
